pingcap/tidb
# util: Automatically create TLS certificates

### What problem does this PR solve?

Problem Summary:

Configuring TLS should be as easy as possible and having to use `mysql_ssl_rsa_setup` from MySQL server should not be needed for simple setups.

### What is changed and how it works?

What's Changed:

If no `ssl-cert` or `ssl-key` are specified: Create a self signed
cert in the temp storage and use that.

This allows TLS to be used when no user created certificates are
available.

Especially for `tiup playground` and other simple cases this should be
sufficient.

Note that for `caching_sha2_password` support we will either need TLS
connections or RSA keypairs. This brings us a step closer in that
direction.


### Related
- "Automatic SSL and RSA File Generation" on https://dev.mysql.com/doc/refman/8.0/en/creating-ssl-rsa-files-using-mysql.html
- https://docs.pingcap.com/tidb/stable/enable-tls-between-clients-and-servers
- https://github.com/pingcap/tidb/issues/9411
- https://github.com/pingcap/tidb/issues/18084



### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Run `tidb-server` without TLS configured and then connect with a modern MySQL client and use `\s` to see if TLS is used.

### Release note 

- A TLS Certificate and Key are automatically generated when TiDB is started without a `ssl-cert` or `ssl-key` configured.
